### ISSUE TITLE
`denote-file-prompt` does not seem to take `denote-file-history` into account

### ISSUE THREAD

**Issue by brabalan** [protesilaos/denote#339]
When I call a command that uses `denote-file-prompt`, I’m surprised that the history is not taken into account. Even though `denote-file-history` is not empty, the first file suggested is always the same one.

Suspicion: the call to `(completing-read prompt new-collection nil nil nil 'denote-file-history)` seems to use a `new-collection` with short (relative) names, whereas the history contains absolute file names. Could it be the reason why it’s not working?

**Comment by protesilaos** [comment]
You mean that `M-p` in the minibuffer does not work? It seems correct on my end, but maybe I am missing something.

**Comment by brabalan** [comment]
The problem may be that I’m using vertico…

When I for instance run `denote-open-or-create` I see this

<img width="882" alt="Screenshot 2024-05-07 at 14 02 44" src="https://github.com/protesilaos/denote/assets/1021637/e821007b-9ee7-4ad4-9ca6-6385eb63c3f2">

and this is unrelated to the history of files I accessed.

If I use `M-p` I indeed have the absolute file name in the prompt

<img width="971" alt="Screenshot 2024-05-07 at 14 05 27" src="https://github.com/protesilaos/denote/assets/1021637/3b9d6ff5-fb82-469e-bd4e-619237757fe1">

but it does not select the file (as the absolute name does not correspond to anything, since it mentions the path), and then I cannot access the note and denote proposes to create it.

Disabling vertico "solves" the issue in the sense that nothing is suggested, and `M-p` does end up finding the note. So I guess this bug report is more about the interaction between vertico and `denote-prompt`.

**Comment by protesilaos** [comment]
I think the issue is with our prompt and how the history is recorded. How about this?

```diff
 denote.el | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/denote.el b/denote.el
index a500437..4035d2b 100644
--- a/denote.el
+++ b/denote.el
@@ -1196,12 +1196,10 @@ (defun denote-file-prompt (&optional files-matching-regexp prompt-text)
            (_ (when included-cpd
                 (setq substrings (cons "./" substrings))))
            (new-collection (denote--completion-table 'file substrings))
-           (relname (completing-read prompt new-collection nil nil nil 'denote-file-history))
-           (absname (expand-file-name relname common-parent-directory)))
-      ;; NOTE 2024-02-29: This delete and add feels awkward.  I wish
-      ;; we could tell `completing-read' to just leave this up to us.
-      (setq denote-file-history (delete relname denote-file-history))
-      (add-to-history 'denote-file-history absname)
+           (input (completing-read prompt new-collection nil nil nil nil))
+           (absname (expand-file-name input common-parent-directory)))
+      (setq denote-file-history (delete input denote-file-history))
+      (add-to-history 'denote-file-history (file-name-nondirectory absname))
       absname)))
 
 ;;;; Keywords
```

**Comment by protesilaos** [comment]
I just pushed a change. It is an improved version of the diff I sent earlier. We actually do not want to do anything with the absolute file path.

Please try the update when you can. Thank you!

**Comment by protesilaos** [comment]
I cannot reproduce this. I even did it with `emacs -Q` from the command-line and then evaluated this to run the command:

```elisp
(require 'denote "/home/prot/Git/Projects/denote/denote.el")
```

**Comment by protesilaos** [comment]
Assuming there is some issue there, perhaps it is best we avoid dealing with the history altogether. I remember it causing us troubles before.

Here is a possible alternative:

```diff
 denote.el | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/denote.el b/denote.el
index 0c01dbf..1217b4f 100644
--- a/denote.el
+++ b/denote.el
@@ -1169,6 +1169,11 @@ (defvar denote-file-history nil
 (defalias 'denote--file-history 'denote-file-history
   "Compatibility alias for `denote-file-history'.")
 
+(defvar denote-file-prompt-latest-input nil
+  "Latest input passed to `denote-file-prompt'.
+This is used for retrieving a value that is used to set a new default at
+the title prompt of `denote-open-or-create' and related commands.")
+
 ;; NOTE 2024-02-29: Based on `project--read-file-cpd-relative' from
 ;; the built-in project.el
 (defun denote-file-prompt (&optional files-matching-regexp prompt-text)
@@ -1203,6 +1208,7 @@ (defun denote-file-prompt (&optional files-matching-regexp prompt-text)
                        (insert input)
                        (completion-in-region (point-min) (point-max) new-collection)
                        (buffer-string))))
+      (setq denote-file-prompt-latest-input input)
       (when filename
         (setq denote-file-history (delete input denote-file-history))
         (add-to-history 'denote-file-history filename))
@@ -2105,8 +2111,7 @@ (defun denote--command-with-features (command force-use-file-prompt-as-default-t
          (or force-ignore-region denote-ignore-region-in-denote-command))
         (denote-title-prompt-current-default
          (if force-use-file-prompt-as-default-title
-             (when denote-file-history
-               (file-name-nondirectory (pop denote-file-history)))
+             denote-title-prompt-current-default
            denote-title-prompt-current-default))
         (path))
     (if in-background
@@ -2798,7 +2803,8 @@ (defun denote--rename-get-file-info-from-prompts-or-existing (file)
                            signature
                            (format "Rename `%s' with SIGNATURE (empty to remove)" file-in-prompt))))
          ('file-type
-          (setq file-type (denote--valid-file-type (denote-file-type-prompt))))
+          (when (denote-file-is-note-p file)
+            (setq file-type (denote--valid-file-type (denote-file-type-prompt)))))
          ('subdirectory
           (setq directory (file-name-as-directory (denote-subdirectory-prompt))))
          ('date
```

**Comment by brabalan** [comment]
I cannot reproduce the problem I raised, and I did not change anything.

There is a noticeable delay (of about one second) between the moment I select a match and opening the file (about one second), and I see "Sole completion" echoed in the minibuffer. I searched the denote source and it does not seem to come from there…

So to sum up, I think the version that you recently pushed works, and there seems to be issues on my end. Thanks again and sorry for the trouble.

**Comment by protesilaos** [comment]
Strange behaviour... Did this happen with my latest suggested diff or without it? I think the approach of that diff is safer for us than relying on the history.

**Comment by protesilaos** [comment]
Please check commit `ff7982f` which includes an update for `denote-file-prompt`. The previous diff is outdated.

**Comment by jeanphilippegg** [comment]
**Whatever happens inside `denote-file-prompt`, it is important that it
returns the absolute file path.**

The current `denote-file-prompt` is problematic because it does not return an
absolute file path anymore as we expect everywhere. This is going to cause
issues.

One of brabalan's issues above is caused by this. If you are anywhere else but
`denote-directory` and you call `denote-open-or-create` and you select an
existing file to open, the file will not be opened and you will instead fall
into creation mode. That is because the relative file does not exist from your
current location.

# Context

Here is some historical info about `denote-file-prompt`:

In issue #233 (and #239), a user wanted to display relative file paths in
`denote-file-prompt`. There were a few discussions and we wanted to accomplish
this only visually (ie keep working with absolute file paths underneath).
`denote-file-prompt` would always return an absolute file path.
`denote-file-history` would also always contain absolute paths.

We were not successful at making `denote-file-prompt` display relative paths
and keep everything in absolute form underneath. So we ended up building the
"completion list" (relative paths) and the "history list" (absolute paths)
manually. However, as we can now see, this is causing issues as well, because
the history never seems to affect the completions. This is made visible if you
use Vertico. The "completions list" is displayed and you can access the
"history list" with `M-p`.

# Problem

The issue is likely as Brabalan said: we use relative paths for the
"completions list" and absolute paths for the "history list" and so one does
not seem to affect the other.

What we see in the "completions list" is built using the list that we give to
`completing-read` and the items of the history, but only those that are all in
the "completions list".

# Solution

If we want to keep displaying relative file paths in `denote-file-prompt` and
have a usable history, we seem to be forced to have them both use relative
file paths. And `denote-file-prompt` should always still return an absolute
path. But I also think that the new `denote-file-prompt-latest-input` is good,
so let's keep it!

I have made the changes in pull request #341. Please checkout the latest code
and evaluate the code below and check if it fixes your issue. Your history
will be rebuilt over time with relative paths.

```emacs-lisp
(defun denote-file-prompt (&optional files-matching-regexp prompt-text)
  (let* ((relative-files (mapcar #'denote-get-file-name-relative-to-denote-directory
                                 (denote-directory-files files-matching-regexp :omit-current)))
         (input (completing-read
                 (or prompt-text "Select FILE: ")
                 (denote--completion-table 'file relative-files)
                 nil nil nil 'denote-file-history)))
    (setq denote-file-prompt-latest-input input)
    (concat (denote-directory) input)))
```

I was able to reproduce the issues discussed in this issue:
- `denote-file-history` not considered using the previous code
- Creation mode when selecting an existing note in `denote-open-or-create`.
- Delay of 1 second, then "Sole completion" with current code.

The code above seems to fix them.

**Comment by protesilaos** [comment]
Thank you @jeanphilippegg!

We still need to account for more cases. Here is the proposed change on top of yours:

```diff
 denote.el | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/denote.el b/denote.el
index 63aa582..44ed125 100644
--- a/denote.el
+++ b/denote.el
@@ -1189,9 +1189,17 @@ (defun denote-file-prompt (&optional files-matching-regexp prompt-text)
          (input (completing-read
                  prompt
                  (denote--completion-table 'file relative-files)
-                 nil nil nil 'denote-file-history)))
-    (setq denote-file-prompt-latest-input input)
-    (concat (denote-directory) input)))
+                 nil nil nil 'denote-file-history))
+         (full-candidate (with-temp-buffer
+                           (insert input)
+                           (completion-in-region (point-min) (point-max) relative-files)
+                           (buffer-string))))
+    (if-let ((file (expand-file-name full-candidate (denote-directory)))
+             ((file-exists-p file)))
+          file
+      (setq denote-file-history (delete input denote-file-history))
+      (setq denote-file-prompt-latest-input input)
+      input)))
 
 ;;;; Keywords
 

```

The reason for this is that our `completing-read` does not have a `REQUIRE-MATCH` parameter. We have it without one to facilitate the "... or act" type of commands we have. Without this requirement, anything we input is accepted as-is and is then recorded in the history of files even though it may not necessarily match any file.

Storing the last non-file input in the hope that it will be reified as a file later (e.g. `denote-open-or-create`) is also not reliable, as the user may change their mind at the following title prompt.

The goal is to have historical inputs match file names, so cycling through them in the minibuffer will return some candidates.

We could have a separate history for non-matching inputs, though I think we do not need to keep that as such a history would not be useful in any prompt. Hence the `denote-file-prompt-latest-input`.

Note that I am testing this without Vertico, which allows me to hit `RET` without it selecting the first candidate. Try to input something that matches a file and then with `M-RET` with Vertico to see how it accepts the input literally.

**Comment by jeanphilippegg** [comment]
Ok, I understand that we need to keep the history clean (and only containing
existing files).

I am not sure I understand the use of `completion-in-region`. What I get with
your lastest diff is this: If I type "200" (not an existing file), then we
will autocomplete to "/home/user/notes/20090101T010101--note.txt". I think
maybe it would be better to use "200" literally, instead of autocompleting it
ourselves.

Also your patch returns `input`, which is relative. This may still cause
issues depending on where `denote-open-or-create` is called.

What about the following version?

```emacs-lisp
(defun denote-file-prompt (&optional files-matching-regexp prompt-text)
  (let* ((relative-files (mapcar #'denote-get-file-name-relative-to-denote-directory
                                 (denote-directory-files files-matching-regexp :omit-current)))
         (prompt (format "%s in %s: " (or prompt-text "Select FILE") (denote-directory)))
         (input (completing-read
                 prompt
                 (denote--completion-table 'file relative-files)
                 nil nil nil 'denote-file-history))
         (absolute-file (concat (denote-directory) input)))
    (unless (file-exists-p absolute-file)
      (setq denote-file-history (delete input denote-file-history))
      (setq denote-file-prompt-latest-input input))
    absolute-file))
```

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Fri, 10 May 2024 20:41:35 -0700

> [... 4 lines elided]

> I am not sure I understand the use of `completion-in-region`. What I get with
> your lastest diff is this: If I type "200" (not an existing file), then we
> will autocomplete to "/home/user/notes/20090101T010101--note.txt". I think
> maybe it would be better to use "200" literally, instead of autocompleting it
> ourselves.

The completion is done to see if an incomplete input is matching any
file. For example, I type "something-unique RET" and want to exit with
the matching file.

With your new suggestion the value of 'absolute-file' will be something
like "path/to/something-unique" instead of the actual file.

This is because we are not using the REQUIRE-MATCH of 'completing-read',
which would force the minibuffer to exit with the expanded candidate
instead of the literal input.

We want the literal input only if there is no matching file.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by brabalan** [comment]
Thanks a lot for the insightful discussion. I’ll test this when I get some time next week.

**Comment by jeanphilippegg** [comment]
> The completion is done to see if an incomplete input is matching any file.
  For example, I type "something-unique RET" and want to exit with the matching file.

Sorry, I still don't get it. Shouldn't we let the user manage the completion
here with `tab`? Otherwise, if the user types "something-unique", it is not
wrong for us to assume that he wants to create a new note named
"something-unique"?

> With your new suggestion the value of 'absolute-file' will be something
like "path/to/something-unique" instead of the actual file. This is because we
are not using the REQUIRE-MATCH of 'completing-read', which would force
the minibuffer to exit with the expanded candidate instead of the literal input.
We want the literal input only if there is no matching file.

If the input is "something" (non existant) and "/path/to/notes/something" does
not exist as a file, we do not really want to **return** the litteral input
"something". We actually need to put "something" in
`denote-file-prompt-latest-input` and then return "/path/to/notes/something"
(non existant).

This is because the users of `denote-open-or-create` expect an absolute path
and will handle this non existant file appropriately. Otherwise, if we return
"something", this detection by `denote-open-or-create` may fail depending on
the location of the call. For example, if there is a file named "something" in
"/home" and we call it from there, `denote-open-or-create` will assume that
the file "something" exists and not try to create a new note.

In short, we need to return "/path/to/notes/something" wether it exists or
not. And we put "something" in `denote-file-prompt-latest-input`.

**Comment by protesilaos** [comment]
> Sorry, I still don't get it. Shouldn't we let the user manage the completion
here with tab? Otherwise, if the user types "something-unique", it is not
wrong for us to assume that he wants to create a new note named
"something-unique"?

I think we still need to try to match the candidate. Because if the user does this in other prompts, they do get a match. With Vertico (a relevant package given its popularity), this is the default behaviour. We even have notes in our documentation about how to make it accept a literal empty string.

Our completion table here adds the `file` category which means that we make this prompt subject to the user option `vertico-preselect`. The user can set this to focus on the prompt (which is useful with `find-ile`, Dired renaming, etc.). When this happens, the user cannot exit with an existing match by typing just a substring. They must use TAB, which may not be what they normally do.

With or without Vertico, a `completing-read` will exit with the unique match of a substring if `REQUIRE-MATCH` is passed to it. Many of our prompts are this way. Those that do not have this argument are designed to accept an empty string.

If the user really wants to create a new note no matter what, they then should use `denote` (or related directly). Here we have to have a more involved approach.

> If the input is "something" (non existant) and "/path/to/notes/something" does
not exist as a file, we do not really want to return the litteral input
"something". We actually need to put "something" in
denote-file-prompt-latest-input and then return "/path/to/notes/something"
(non existant).

Yes, this is better.

* * *

The above clarify what I think should happen, but please do not take this as a directive. I am happy to do what you or others think should be happening. I am especially interested in feedback from people who use `denote-open-or-create` regularly (like @brabalan who reported this): I only use `denote-open-or-create` for testing purposes.

* * *

On another note, perhaps it is time we decouple the "... or create" logic from the file prompt, so that the latter would do only one thing: return a file or nil. Then the caller would take care to handle the nil case. Any thoughts?

**Comment by jeanphilippegg** [comment]
> With or without Vertico, a `completing-read` will exit with the unique match of a substring if `REQUIRE-MATCH` is passed to it. Many of our prompts are this way. Those that do not have this argument are designed to accept an empty string.

Ok, I understand now. Thank you!

> On another note, perhaps it is time we decouple the "... or create" logic from the file prompt, so that the latter would do only one thing: return a file or nil. Then the caller would take care to handle the nil case. Any thoughts?

What do you have in mind for when a user calls `denote-open-or-create`? There has to be some kind of prompt that detects whether the input is an existing note or a new one.

Until we figure this out, do you want me to make a pull request to fix the current `denote-file-prompt`? In the end, the only change will have been that the "history list" is now relative instead of absolute, thus fixing the issue in the original post.

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Sat, 11 May 2024 08:52:33 -0700
>
>> With or without Vertico, a `completing-read` will exit with the
>> unique match of a substring if `REQUIRE-MATCH` is passed to it. Many
>> of our prompts are this way. Those that do not have this argument are
>> designed to accept an empty string.
>
> Ok, I understand now. Thank you!

You are welcome!

>> On another note, perhaps it is time we decouple the "... or create"
>> logic from the file prompt, so that the latter would do only one
>> thing: return a file or nil. Then the caller would take care to
>> handle the nil case. Any thoughts?
>
> What do you have in mind for when a user calls
> `denote-open-or-create`? There has to be some kind of prompt that
> detects whether the input is an existing note or a new one.

The functionality should look the same for the user, though internally
we can have a wrapper function that is specific for this "... or create"
logic. It will test if denote-file-prompt returns nil (or no existing
file---whatever works better), at which point it will create the new
one. This wrapper function will store the user's input as the value of
denote-file-prompt-latest-input and can even have its own separate
minibuffer history (depending on how we choose to design it).

The goal is to reduce denote-file-prompt to a prompt like all others. We
should even set the REQUIRE-MATCH to non-nil and arrange to disable it
only on demand, such as for this special case. Otherwise, we do want to
match a candidate when a substring uniquely refers to it (the issue I
described before happens with denote-link, for example, which is not a
good experience for such a common workflow).

> Until we figure this out, do you want me to make a pull request to fix
> the current `denote-file-prompt`? In the end, the only change will
> have been that the "history list" is now relative instead of absolute,
> thus fixing the issue in the original post.

Yes, please go ahead.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by jeanphilippegg** [comment]
Ok, I think I understand the subtleties, but you probably have a better idea of the contexts in which we want to disable REQUIRE-MATCH. Is that ok if let you go ahead with this change? It seems good to me!

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Sat, 11 May 2024 10:33:34 -0700
>
> Ok, I think I understand the subtleties, but you probably have a
> better idea of the contexts in which we want to disable REQUIRE-MATCH.
> Is that ok if let you go ahead with this change? It seems good to me!

Sure. I will make some changes tomorrow and then we can review them.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by protesilaos** [comment]
I just committed the changes. I think now the intent is clear. If there are any further refinements, we continue with them. Please let me know.

**Comment by brabalan** [comment]
I tested these changes and they work great for me. Thanks again!

**Comment by jeanphilippegg** [comment]
This is still problematic. :/

`denote-file-prompt` needs to always return an absolute path. The explanation is in my comment above:

> If the input is "something" (non existant) and "/path/to/notes/something" does not exist as a file, we do not really want to **return** the litteral input "something". We actually need to put "something" in `denote-file-prompt-latest-input` and then return "/path/to/notes/something" (non existant).
> 
> This is because the users of `denote-open-or-create` expect an absolute path and will handle this non existant file appropriately. Otherwise, if we return "something", this detection by `denote-open-or-create` may fail depending on the location of the call. For example, if there is a file named "something" in "/home" and we call it from there, `denote-open-or-create` will assume that the file "something" exists and not try to create a new note.
> 
> In short, we need to return "/path/to/notes/something" whether it exists or not. And we put "something" in `denote-file-prompt-latest-input`.

**Comment by jeanphilippegg** [comment]
In particular, `denote-file-or-action-prompt` also needs to return an absolute path for the same reason.

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Sun, 12 May 2024 06:58:02 -0700
>
> This is still problematic. :/
>
> `denote-file-prompt` needs to always return an absolute path. The explanation is in my comment above:
>
>> If the input is "something" (non existant) and
>> "/path/to/notes/something" does not exist as a file, we do not really
>> want to **return** the litteral input "something". We actually need
>> to put "something" in `denote-file-prompt-latest-input` and then
>> return "/path/to/notes/something" (non existant).
>> 
>> This is because the users of `denote-open-or-create` expect an
>> absolute path and will handle this non existant file appropriately.
>> Otherwise, if we return "something", this detection by
>> `denote-open-or-create` may fail depending on the location of the
>> call. For example, if there is a file named "something" in "/home"
>> and we call it from there, `denote-open-or-create` will assume that
>> the file "something" exists and not try to create a new note.
>> 
>> In short, we need to return "/path/to/notes/something" whether it
>> exists or not. And we put "something" in
>> `denote-file-prompt-latest-input`.

Fine, I did not account for this case. Do you want to send a PR for
this?

What about the rest of the functionality?

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by jeanphilippegg** [comment]
Sure! I made pull request #348.

I think the rest of the functionality is good. Though, I think we may still
want to use `denote-file-history` even in `denote-open-or-create` to have
candidates to choose from.

Even though it is not necessary, I made it back to a single function (but
keeping the `no-require-match` parameter) for these reasons:

- If I make it return an absolute path, it would not be doing much. It would
  only call `denote-file-prompt` with `no-require-match` set to `t` (and set
  last-input).
- Soon, I will have to arrange `denote-file-prompt` and
  `denote-select-linked-file-prompt` because `denote-directory` will accept a
  list of directories. It will be simpler to handle 2 functions instead of 3.

Let me know if you would rather keep the separate
`denote-file-or-action-prompt`. We can still have that.

**Comment by protesilaos** [comment]
Thank you @jeanphilippegg! I merged your PR. I don't think we need a separate prompt, especially since you plan to build on the one we have. The only value of the separate prompt is that it makes the intent/use-case more explicit. Though this is not a big deal: we can always have documentation for that if we feel it is necessary.

@brabalan Please test again when you have the time.

**Comment by brabalan** [comment]
I did a quick test and everything seems to work. I’ll report if I see anything amiss. For me, this issue can be closed. And thanks a lot to both of you!

**Comment by protesilaos** [comment]
Thank you! We can safely close this now.

**Comment by brabalan** [comment]
The problem seems to happen again: for some reason, `denote-file-prompt` seems to ignore `denote-file-history` and suggest files ordered by most recently modified. The same problem seems to happen when I use `consult-file-prompt`.

I use `vertico` with `vertico-sort-function` set to `vertico-sort-history-length-alpha`.

Should I submit a new issue for this, or do you prefer to reopen it?

**Comment by protesilaos** [comment]
Hello @brabalan!

> `denote-file-prompt` seems to ignore `denote-file-history` and suggest files ordered by most recently modified.

This is by design now, because of this:

```elisp
(defvar denote-file-prompt-extra-metadata
  (list
   ;; NOTE 2025-12-15: If we use the `file' category, then we are
   ;; subject to the `completion-category-overrides'.  This is a
   ;; problem because the user will want to, for example, sort
   ;; directories before files, but then we cannot have our sort here.
   (cons 'category 'denote-file)
   (cons 'group-function #'denote-file-prompt-group)
   (cons 'affixation-function #'denote-file-prompt-affixate)
   (cons 'display-sort-function #'denote-file-prompt-sort))
  "Extra `completion-metadata' for the `denote-file-prompt'.
This is in addition to the completion category, which is constant.")
```

[ I now realise that the docstring is slightly wrong, but whatever. ]

There is an explicit sorting, the `display-sort-function`, which is handled by `denote-file-prompt-sort`.

You can either set this variable to `nil` or change it so that it does not have this line:

```elisp
(cons 'display-sort-function #'denote-file-prompt-sort))

```

**Comment by brabalan** [comment]
Ah, that solved it. Thanks!

**Comment by protesilaos** [comment]
Very well!
